enhancements/olm-curated-catalog.md

```diff
@@ -0,0 +1,302 @@
+      --permissive              allow registry load errors
+  -t, --tag string              custom tag for container image being built
+```
+
+Examples:
+```sh
+$ opm index filter -t dockerreg.local/catalogs/company-catalog:latest -f ./filter.yaml
+```
+
+or
+```sh
+$ cat filter.yaml | opm index filter -t dockerreg.local/catalogs/company-catalog:latest  -f -
+```
+
+The `filterSpec` yaml specification is as follows:
+
+* `filterSpec` (array of objects): array of filter specification objects as follows
+  * `indexImage` (string): the catalog index image refererence
+  * `filter` (array of objects):  array of filter objects as follows:
+    * `package` (string) : the name of the operator package to include.
+    * `includedChannels` (array of string): the array of channel names to include from the package (Optional, defaults to all channels)
+    * `excludedChannels` (array of string): the array of channels to exclude (Optional, defaults to no excluded channels)
+    
+    This is useful to use to support future channels without updating the filter and takes precedence over the `includedChannels`
+     
+
+ "quay.io/openshift/community-operators-catalog:latest"
+
+Example (in yaml for readabiilty)
+```yaml
+filterSpec:
+- indexImage:  "quay.io/openshift/community-operators-catalog:latest"
+  filter:
+  - package: etcd
+    # Include only the clusterwide channels.
+    includedChannels:
+    - beta-clusterwide
+    - stable-clusterwide
+  - package: postresql
+    # Include only the stable channel.
+    includedChannels:
+    - stable
+- indexImage:  "mydockerreg.local/myapp-catalog:latest"
+  - package: myapp1
+    # No includedChannels or excludedChannels, implies ALL channels
+  - package: myapp2
+    includedChannels:
+    # Include any 1.1.x and 2.x operators
+    - v1.1
+    - v2
+  - package: myapp3
+    # Include every channel except for the 1.x and 2.x beta channels
+    excludedChannels:
+    - v1
+    - v2-alpha
+    - v2-beta
+```
+#### Filter Flow
+`opm index filter` will be added to perform the equivalent steps (non-optimized):
+
+- For each `filterSpec` as `FSPEC`:
+  - Extract the source `FSPEC.indexImage` into a temp file structure as follows:
+    ```
+    catalog
+      package
+        bundle
+          metadata/annotations.yaml (with channels info)
+          ...
+    ```
+  - Delete any packages and bundles that don't match the filter:
+
+    For each extracted `catalog.package` as `PKG`:
+    - if `PKG` NOT IN `FSPEC.filter.packages`, delete the package
+    - else: 
+      - for each `PKG.bundle.channels` AS `CHN`
+        - if `CHN` IN `FSPEC.filter.PKG.excludedChannels` OR (`FSPEC.filter.PKG.includedChannels` != Empty AND `CHN` NOT IN `FSPEC.filter.PKG.includedChannels`)
+          - delete `PKG.bundle`
+      - Create the bundle image with only the intersecting `PKG.bundle.channels` and the filtered channels as `NEWBNDL`
```

Comment:
Instead of this, could we "save" the filter spec (and a potential status, maybe that includes the digest of the index image used) in the index image. (for reproducibility) and then we manipulate the database instead of rebuilding bundles?

/cc @kevinrizza @ecordell

Reply:
Yeah that seems like the right way to do this. I don't think filter should be recreating the index, rather we should just modify the database in the same way that `index add` or `index remove` are already built.

Reply:
My proposal was assuming that the Bundle maintained the source of truth.  If we say that catalog has as filtered opinion of the original bundles, that's fine with me.   

My pseudo-code was mostly for illustration to describe how it would work from the external API perspective.

Because I'm not an expert in the code itself, I'm hesitant to make any assumptions.  Especially since I'm not 100% sure on how the graph is supposed to work due to various bugs I'm finding in `opm`.